### ISSUE TITLE
Improve support for the rr debugger in nixos containers

### ISSUE THREAD

**Issue by deliciouslytyped** [NixOS/nixpkgs#99535]
Several factors contibuted to having some difficulty with using rr ( https://github.com/mozilla/rr ) in a nixos container (which is based on systemd-nspawn, I use https://github.com/erikarvstedt/extra-container) - my ordeal can be observed here: https://github.com/mozilla/rr/issues/2696#issuecomment-702869449 .

The summary is
- rr has not had a release for a long time, during which they have accumulated improvements also related to error checking
- systemd-nspawn makes it difficult to fully disable seccomp

I think the necessary steps to help here is
- [ ] bump the rr package version to the tip of the master branch, at least until there is a new release
- [ ] create an rr related config option (restricted to containers?) that adds the necessary systemd-nspawn seccomp permissions

cc @thoughtpolice @nbp because maintainers

**Comment by deliciouslytyped** [comment]
https://github.com/gsliepen/tinc/issues/252#issue-705902018 should contain all the information I needed to get seccomp working, albeit we would want to do things cleaner;*
https://github.com/mozilla/rr/wiki/Docker has a little bit of information about what rr needs enabled. but it mostly just disables all restrictions
- [ ] it would probably be nice if there was a wiki article with more details, or at least som manner of documentation to refer to

`--system-call-filter` can be used to whitelist (more info in tinc issue) syscalls, the implementation can be found at: TODO
either:
- [ ] figure out if a wildcard can be used to whitelist everything
- [ ] manually whitelist individual items
- [ ] It may be the case that only exposes a means to whitelist named syscalls; rr also has a specific extra numeric range which it uses for internal communication (TODO: xlink). Lacking this, users must use the `-n` flag with `rr record` (sufficiently new versions tell you this when you attempt to start it).

 *I didn't actually go through with it all the way and just used `-n`

**Comment by deliciouslytyped** [comment]
The apppropriate section of the systemd v243 manual states: https://github.com/systemd/systemd/blob/efb536d0cbe2e58f80e501d19999928c75e08f6a/man/systemd-nspawn.xml#L979-L994

**Comment by deliciouslytyped** [comment]
Syscall filter set definitions, including "@default" can be found at https://github.com/systemd/systemd/blob/efb536d0cbe2e58f80e501d19999928c75e08f6a/src/shared/seccomp-util.c#L255 . 

The name resolution functionality appears to be implemented seccomp_parse_syscall_filter in seccomp-util.c via https://man7.org/linux/man-pages/man3/seccomp_syscall_resolve_name.3.html m which doesn't seem to resolve things like "1000".

This means whitelisting rr syscall numbers is is impossible, and this issue is blocked on one of:
- systemd implements unfiltered nspawn
- systemd implements syscall number parsing
- seccomp implements accepting integers as strings

**Comment by deliciouslytyped** [comment]
Some additional clarification: per https://github.com/mozilla/rr/issues/2696#issuecomment-702430986 and the following few posts, it seems to me that rr would not have it's syscall names in whatever is used by seccomp to resolve names to numbers.

**Comment by deliciouslytyped** [comment]
@khuey is there an enumaration of the syscalls rr needs, or should one expect to enable every syscall?
It might be nice to tone down the shotgun approach a little.

**Comment by deliciouslytyped** [comment]
One can use the `ausyscall` program from the `audit` package to enumerate syscalls.

@khuey should programs run in `rr record` be able to see rr's syscalls? ausyscall doesnt list any syscalls around 1000 in `rr record ausyscall --dump`. To be clear, I don't expect it to, I'm just checking.

Edit: ok, jury is still out on syscall dump tools. Most tools just seem to dump from a precompiled table.

**Comment by khuey** [comment]
I can see why it would be confusing, but the section about seccomp in the rr technical report is largely irrelevant here. rr has two recording modes, with and without "syscall buffering" (the default and what you get with `record -n`, respectively). Without syscall buffering, rr relies entirely on ptrace to observe the behavior of the tracee process (tree). With syscall buffering, rr uses a library (librrpreload.so) inserted into the tracee process (tree) via LD_PRELOAD to observe certain common and fast syscalls in process without switching to the rr tracer process, and falls back on ptrace to observe other complex, slow, uncommon, etc syscalls. A seccomp filter is installed by rr on the tracee process (tree) to choose which syscalls trigger ptrace traps and which are handled in process. This is what the seccomp section of the rr technical report is about.

librrpreload needs to communicate with the rr tracer process sometimes. To do this, because rr is already observing the tracee process (tree) via ptrace, librrpreload simply makes syscalls with special private syscall numbers. The rr tracer sees these syscalls before the kernel attempts to act on them (and realizes they don't exist and returns ENOSYS) and handles them appropriately. If a seccomp filter is installed by a third party (such as nixos, docker, etc) on *rr* (really rr's tracees, but distinguishing between rr itself and the tracee process (tree) would be difficult to impossible) and it is does not allow these private syscalls to proceed then syscall buffering will not work. This is what you ran into in the issue in the rr repository we worked through before.

In practice that means any external seccomp filter applied to rr and its tracees needs to be default-allow or explicitly allow the private syscall range we use (which is currently 1000-1009). We have not previously attempted to catalog the full list of syscalls required for rr because a number of the syscalls we use are forbidden by every sandboxing seccomp filter I've ever seen (including ptrace, seccomp, and process_vm_writev) and because rr is a debugging tool so simply disabling a seccomp filter is usually the easiest course of action (as you've seen, this is our recommended approach for docker).

However apparently systemd-nspawn doesn't allow that so here we are.

**Comment by khuey** [comment]
> @khuey should programs run in `rr record` be able to see rr's syscalls? ausyscall doesnt list any syscalls around 1000 in `rr record ausyscall --dump`. To be clear, I don't expect it to, I'm just checking.

In general no, a tracee would never see rr's private syscalls. They are listed in the `rr dump` output (as `rrcall_whatever`)

**Comment by deliciouslytyped** [comment]
Systemd seems to have given the OK on implementing the necessary things, now I just need to find someone willing to implement it. It doesn't seem difficult (famous last words) but I'm not comfortable writing C.

**Comment by BK1603** [comment]
I'd like to work on the systemd issue for this feature. 
(I should probably mention this on a comment on the systemd issue too.)

**Comment by deliciouslytyped** [comment]
Note rr has made a new release. 
https://github.com/mozilla/rr/releases/tag/5.4.0

**Comment by khuey** [comment]
> Note rr has made a new release.
> https://github.com/mozilla/rr/releases/tag/5.4.0

Beyond including a more useful error message for the case you hit, there shouldn't be relevant changes for you.

**Comment by deliciouslytyped** [comment]
Yeah I made a note because 1) I didn't check if we updated the "stable" package yet, 2) I added an "unstable" attribute to temporarily hold a package with the fixes, so (2) is now unnecessary.
